### DISCUSSION TITLE
How to override default keycloak message with my own one

### DISCUSSION THREAD

**Discussion by LoVas2** [keycloakify/keycloakify#84] [Q&A]
I'd like to override the default `invalidUserMessage` by my own one. I did the following in a `KcMessages.tsx` : 
```typescript
import { kcMessages } from "keycloakify/lib/i18n/useKcMessage";

Object.assign(
	kcMessages["fr"],
	{
		/* spell-checker: disable */
		"invalidUserMessage": "foo en Francais",
		/* spell-checker: enable */
	}
);

Object.assign(
	kcMessages["en"],
	{
		"invalidUserMessage": "foo in English",
	}
);
```
Then, I imported this file in my `CardComponent.tsx` where I make routing thanks to my `KcContext` :
```typescript
import "../model/KcMessages";

export default class CardComponent extends Component<CardProps, CardState> {

    getKcComponent(kcContext: KcContext) {
        switch(kcContext.pageId) {
            case 'login.ftl':
                return <Login kcContext={kcContext} {...{...kcProps,}} />;
            case 'login-update-password.ftl':
                return <UpdatePassword kcContext={kcContext} {...{...kcProps,}} />;
            case 'info.ftl':
                return <Info kcContext={kcContext} {...{...kcProps,}} />;
            default:
                return <KcAppBase kcContext={kcContext} {...{...kcProps,}} />;
        }
    }

    render() {
        return (
                {(kcContext !== undefined) ? this.getKcComponent(kcContext) : <BrowserRouter>...</BrowserRouter>}
        )
    }
```
What did I miss ?

**Comment by garronej** [comment]
Hi @LoVas2,  
I don't know. It looks good to me, it should work....

**Comment by LoVas2** [comment]
I tried on the project `keycloakify-demo-app` branch `look_and_feel` and the message_*.properties still not appear in the compiled jar 😞

**Reply by garronej** [reply]
What are you expecting exactly?  
What are you doing exactly that makes you think it doesn't work?  
I am currently doing it [this way](https://github.com/etalab/sill-web/blob/6268b67949a5b377363fb9607accb9f1e35d6569/src/ui/components/KcApp/kcContext.ts#L94-L127) and it does work...

**Reply by garronej** [reply]
It does work as expected...

https://user-images.githubusercontent.com/6702424/160149279-a1b25ea1-8c69-4454-89f1-c5e2a0940466.mov

**Comment by LoVas2** [comment]
Hello,

Yes it’s working like this. But it’s doesn’t correspond to my needs. 

What I want : override alert message coming from Keycloak server.
When I enter a wrong login or password for example, the Keycloak server send a MessageBean containing a type and a summary. This summary comes from Message Bundle which is using i18n properties files for translation using an error code. If we want, we can override this message summary by providing a custom message_*.properties at the same level that the /ressources folder in the login directory. 

<img width="373" alt="Capture d’écran 2022-03-28 à 09 59 31" src="https://user-images.githubusercontent.com/8726977/160356538-4559a8cd-585a-46e0-8ce1-7e427230891c.png">
<img width="588" alt="Capture d’écran 2022-03-28 à 10 20 20" src="https://user-images.githubusercontent.com/8726977/160356545-77f7d2c1-0b1c-4280-ab4b-baf290cc5e7c.png">


What I am doing with Keycloakify : providing a custom message for an error code. It’s working when I use the {msg(‘InvalidLoginPassword’)} in the template. But the alert comes with a type and a summary and not with a code, so it is not using my KCMessageExtensions and doesn’t override the summary coming from server. 
```javascript
{displayMessage && message !== undefined && (message.type !== "warning" || !isAppInitiatedAction) && (
    <div className={cx("alert", `alert-${message.type}`)}>
        {message.type === "success" && <span className={cx(props.kcFeedbackSuccessIcon)}></span>}
        {message.type === "warning" && <span className={cx(props.kcFeedbackWarningIcon)}></span>}
        {message.type === "error" && <span className={cx(props.kcFeedbackErrorIcon)}></span>}
        {message.type === "info" && <span className={cx(props.kcFeedbackInfoIcon)}></span>}
        <span
            className="kc-feedback-text"
            dangerouslySetInnerHTML={{ "__html": message.summary, }}
        />
    </div>
)}
```

What I did for making work :
Extending the Keycloak server to provide the error code in the MessageBean : 
- extend MesageBean, add code attribute. 
- Add custom FreemarkerLoginProvider, so when the MessageBean is created translate the code with message bundle + add code in attribute. 
- In Keycloakify : update where the alerts are handled and use the new error code with advanced msgStr(message.code)

```javascript
{message !== undefined && (
    <div className={`alert ${message.type}`}>
        <img id={cx(`alert-img-${message.type}`)}
            alt={`Icône d'alerte ${message.type}`}/>
        <span>{advancedMsg(message.code)}</span>
    </div>
)}
```
